Prototype: array-cell-sensitivity

smowton opened the pull request:
Do not review; this is a test branch for array-cell sensitivity

smowton committed:
Symex: enable array cell sensitivity

smowton committed:
Array cell sensitivity: tolerate zero-length arrays

smowton committed:
Resolve array indices before dereferencing

This may lead to application of array-cell-sensitivity (i.e. querying the symbol
some_array[[1]] instead of some_array[some_index]), leading to higher precision.

Some tests must be amended because better constant propagation causes some test behaviour
to change as symex has better knowledge regarding the targets of pointers.

romainbrenguier committed:
Allow trace to assign in two steps

With field sensitivity on arrays, the element can be initialized in two
steps, first the field f then the field array.

romainbrenguier committed:
Set a limit size for array cell propagation

The use of field_sensitivity for arrays can be expensive for big arrays
so we have to set a limit on the size of constant arrays which get
processed.
The value 64 was determined by changing the size of the array in the
test regression/cbmc/variable-access-to-constant-array and reducing it
until there was no significant difference in execution time between the
versions of CBMC with and without array cell propagation.
For the chosen value the execution time was around 0.05 second in both
cases (for comparison with size 1024 the time with propagation was 0.5
sec, against 0.1 without).

romainbrenguier committed:
Adapt double_deref test for array cell sensitivity

Propagatiof of values of array cells lead to different expressions in VCC.

romainbrenguier committed:
Attempt to get missing array size from symbol table

In some cases, like the test in regression/cbmc/Global_Initialization2,
the array type is incomplete and changed in the symbol table after the
main function has been converted, leading to inconsistencies.
This means we can get nil instead of the size in
field_sensitivityt::apply, though the actual size is present in the
symbol table.

romainbrenguier committed:
Make field_sensitivity::apply handle non-constant index

When the index is not constant when accessing a constant size array, we
should expand the expression into `{array[0]; array[1];...}[index]` so
that the relation with field sensitivity expressions `array[[0]]`,
`array[[1]]`, etc, is not lost.

romainbrenguier committed:
Only apply field_sensitivity to dereferenced object

This is important to consider dereferenced object as read value.

romainbrenguier committed:
Apply field sensitivity in symex_printf

This may be necessary for constant propagation of the format string.
An example of that is in regression/cbmc/printf1

romainbrenguier committed:
Tag test failing with SMT backend

This is failing because of this bug:
https://github.com/diffblue/cbmc/issues/4749

romainbrenguier committed:
Rename size to L2 in field_sensitivity::apply

This is to allow constant propagation to take place and know whether the
size is actually constant, which can allow field sensitivity to apply.

romainbrenguier committed:
Simplify lhs in build goto trace

Avoid lhs of the form `{array[0], array[1]}[0]`

smowton committed:
Symex rename functions: use NODISCARD

These used to have side-effects on their parameters, and will still compile if used that
way, so let's use NODISCARD to enforce that their new return value is consulted when
necessary.

smowton committed:
Add tests for array-cell sensitivity

Most are for C, then a couple are duplicated for Java to check whether its use of a structure
to represent all arrays makes any difference.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 64cfecfe74e436d4d0521e05468dba187133df5d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120673312
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton committed:
Add L1_WITH_CONSTANT_PROPAGATION mode to goto_symex_statet::rename

This allows applying constant propagation without upgrading non-constant symbols to L2,
which is useful for the dereferencing logic, as the value-set used to resolve pointer
aliasing is indexed by L1 names.

smowton commented:
Work on this has moved back to https://github.com/diffblue/cbmc/pull/4680
